pydantic/type_adapter.py

```diff
@@ -233,100 +159,203 @@ def __init__(
+        self._init_core_attrs(
+            ns_resolver=_namespace_utils.NsResolver(
+                namespaces_tuple=_namespace_utils.NamespacesTuple(locals=localns, globals=globalns),
+                parent_namespace=localns,
+            ),
+            force=False,
+        )
+
+    def _init_core_attrs(
+        self, ns_resolver: _namespace_utils.NsResolver, force: bool, raise_errors: bool = False
+    ) -> bool:
+        """Initialize the core schema, validator, and serializer for the type.
+
+        Args:
+            ns_resolver: The namespace resolver to use when building the core schema for the adapted type.
+            force: Whether to force the construction of the core schema, validator, and serializer.
+                If `force` is set to `False` and `_defer_build` is `True`, the core schema, validator, and serializer will be set to mocks.
+            raise_errors: Whether to raise errors if initializing any of the core attrs fails.
+
+        Returns:
+            `True` if the core schema, validator, and serializer were successfully initialized, otherwise `False`.
+
+        Raises:
+            PydanticUndefinedAnnotation: If `PydanticUndefinedAnnotation` occurs in`__get_pydantic_core_schema__`
+                and `raise_errors=True`.
+
+        Notes on namespace management, and subtle differences from `BaseModel`:
+            This method is very similar to `_model_construction.complete_model_class`, used for finalizing `BaseModel` subclasses.
+
+            `BaseModel` uses its own `__module__` to find out where it was defined
+            and then looks for symbols to resolve forward references in those globals.
+            On the other hand, `TypeAdapter` can be initialized with arbitrary objects,
+            which may not be types and thus do not have a `__module__` available.
+            So instead we look at the globals in our parent stack frame.
+
+            It is expected that the `ns_resolver` passed to this function will have the correct
+            namespace for the type we're adapting. See `TypeAdapter.__init__` and `TypeAdapter.rebuild`
+            for various ways to construct this namespace.
+
+            This works for the case where this function is called in a module that
+            has the target of forward references in its scope, but
+            does not always work for more complex cases.
+
+            For example, take the following:
+
+            a.py
+            ```python
+            from typing import Dict, List
+
+            IntList = List[int]
+            OuterDict = Dict[str, 'IntList']
+            ```
+
+            b.py
+            ```python test="skip"
+            from a import OuterDict
+
+            from pydantic import TypeAdapter
+
+            IntList = int  # replaces the symbol the forward reference is looking for
+            v = TypeAdapter(OuterDict)
+            v({'x': 1})  # should fail but doesn't
```

Comment:
Were the examples etc here meant to be in public `__init__` side (which has the experimental-looking `_parent_depth` argument for doing the namespace things) for documenting the ForwardRefs quirks?

Reply:
Yeah, that probably makes more sense, given that's where we do the namespace fetching anyways. Happy to move there.

Reply:
Might put this under a closed by default block as to not overwhelm users not using this complex functionality.

Reply:
> Might put this under a closed by default block as to not overwhelm users not using this complex functionality.

Yeah the ForwardRefs are such a can of worms... Not really necessarily relevant for the usual simple use-cases.

Reply:
`rebuild` could be also one place to have this, not sure which one is better...

Reply:
Maybe the class itself, honestly. Bc it applies to both `__init__` and `rebuild`